File: permute/permute.cpp


#include "../include/include.hpp"

class Solution
{
public:
	vector<vector<int>> permute(vector<int> &nums)
	{
		vector<vector<int>> res;

		sort(nums.begin(), nums.end());
		int size = nums.size();

		do
		{
			res.push_back(nums);
		} while (std::next_permutation(nums.begin(), nums.end()));
		return (res);
	}
};

int main()
{
	Solution ss;
	vector<vector<int>> nums = {
	{1, 2, 3},
	{0,-1,1},
	{1},
	{0, 1}
	};

	for (auto num : nums)
	{
		cout << "result for : ";
		print_container(num);
		print_nested_container(ss.permute(num));
		cout << ("-----------------\n");
	}

	return (0);
}

File: permute/permute.py

from typing import List, Optional
import collections



class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []

        def dfs(i, curr, pick):
            if len(curr) == len(nums):
                res.append(curr.copy())
                return
            
            for m in range(len(nums)):
                if not pick[m]:
                    curr.append(nums[m])
                    pick[m] = True
                    dfs(m + 1, curr, pick)
                    curr.pop()
                    pick[m] = False
    
        dfs(0, [], [False] * len(nums))

        return res




nums = [1,2,3]


ss = Solution()

print(ss.permute(nums))




